libyang/data.py

```diff
@@ -224,6 +281,31 @@ def validate(
         if ret != 0:
             raise self.context.error("validation failed")
 
+    def diff(
+        self,
+        other: "DNode",
+        no_siblings: bool = False,
+        with_defaults: bool = False,
+    ) -> Iterator[DDiff]:
+        flags = diff_flags(no_siblings=no_siblings, with_defaults=with_defaults)
+        difflist = lib.lyd_diff(self.cdata, other.cdata, flags)
+
+        i = 0
+        while True:
+            dtype = difflist.type[i]
+            first = None
+            second = None
+            if dtype == lib.LYD_DIFF_END:
+                break
+            if difflist.first[i] != ffi.NULL:
+                first = self.new(self.context, difflist.first[i])
+            if difflist.second[i] != ffi.NULL:
+                second = self.new(self.context, difflist.second[i])
+            yield DDiff(dtype, first, second)
+            i += 1
+
+        lib.lyd_free_diff(difflist)
```

Comment:
This will not be called if there is an error or if the generator is not consumed completely. You should wrap it with a `try`...`finally` block to make sure the allocated diff list is freed.